telosnetwork/teloscan
# #171 | readwrite interface doesnt handle arrays properly

# Fixes #171

## Description

This PR adds in some functionality as a stopgap until function interfaces are overhauled in #104. Currently, several types of ABI function types are not parsed correctly, leading to errors (see Issue for more info)
- adds support for `address`, `address[n?]`, `bool`,  `bool[n?]` , and `string`
- adds support for fixed vs unbounded arrays for above types as well as `uint256`, eg. `uint256[3]` is not equivalent to `uint256`
- adds warning for inputs corresponding to unimplemented types, so the user is aware of current limitations
- adds rudimentary error message & prevents method execution if supported input type has invalid value
- adds hint (in the form of placeholder text) indicating what a correct value should look like

## Test scenarios

- run `yarn dev`
- copy this mock ABI, created to test each type of validated input
    ```json
    [{"inputs":[{"internalType":"string","name":"fakeStringField","type":"string"}],"name":"fakeWriteMethodWithString","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"string","name":"fakeStringField","type":"string"}],"name":"fakeReadMethodWithString","outputs":[],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"address","name":"fakeAddressField","type":"address"}],"name":"fakeReadMethodWithAddress","outputs":[],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"address[]","name":"fakeUnboundedAddressArrayField","type":"address[]"}],"name":"fakeReadMethodWithUnboundedAddressArray","outputs":[],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"address[2]","name":"fakeFixedSizeAddressArrayField","type":"address[2]"}],"name":"fakeReadMethodWithFixedSizeAddressArray","outputs":[],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"bool","name":"fakeBoolField","type":"bool"}],"name":"fakeReadMethodWithBool","outputs":[],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"bool[]","name":"fakeUnboundedBoolArrayField","type":"bool[]"}],"name":"fakeReadMethodWithUnboundedBoolArray","outputs":[],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"bool[3]","name":"fakeFixedSizeBoolArrayField","type":"bool[3]"}],"name":"fakeReadMethodWithFixedSizeBoolArray","outputs":[],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"uint256","name":"fakeUint256Field","type":"uint256"}],"name":"fakeReadMethodWithUint256","outputs":[],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"uint256[]","name":"fakeUnboundedUint256ArrayField","type":"uint256[]"}],"name":"fakeReadMethodWithUnboundedUint256Array","outputs":[],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"uint256[3]","name":"fakeFixedSizeUint256ArrayField","type":"uint256[3]"}],"name":"fakeReadMethodWithFixedSizeUint256Array","outputs":[],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"bytes32","name":"fakeBytes32Field","type":"bytes32"}],"name":"fakeReadMethodWithBytes32","outputs":[],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"uint8[3]","name":"fakeFixedSizeUint8ArrayField","type":"uint8[3]"}],"name":"fakeReadMethodWithFixedSizeUint256Array","outputs":[],"stateMutability":"view","type":"function"}]
    ```
- go to any unvalidated contract address page, eg. http://192.168.236.164:8080/address/0x6ac4c48f04dd498dae1e043da096583c4d4719b3#contract
- click the "ABI from JSON" tab
- paste the mock ABI from above
    - read and write tabs should appear below the ABI field
- for each input in each tab, note the following:
    - the input, if implemented, should display a hint placeholder text when focused, guiding the user as to what a correct value would look like
    - entering any invalid value (including an array of the wrong length for bounded arrays) should result in an error message
    - entering a valid value and executing the method should clear the error message
    - for array types, entering any number of spaces after a comma should not produce an error, eg `[1, 2,      3,4]`
    - if an input is not valid, it should display a warning
